controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,255 @@
+	if isROSAHCP, err := platform.IsPlatformROSAHCP(); err != nil {
+		r.Log.Error(err, "Failed to determine if ROSA HCP cluster")
+		return nil, err
+	} else if isROSAHCP {
+		prometheusURL = fmt.Sprintf("https://prometheus.%s.svc.cluster.local:9339", r.OperatorNamespace)
+	}
+
+	caCertPath := "/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt"
+	tokenPath := "/var/run/secrets/kubernetes.io/serviceaccount/token"
+
+	// create auth roundtripper
+	tlsConfig, err := prometheusconfig.NewTLSConfig(&prometheusconfig.TLSConfig{
+		CAFile: caCertPath,
+	})
+	if err != nil {
+		r.Log.Error(err, "failed to create tls config")
+		return nil, err
+	}
+	settings := prometheusconfig.TLSRoundTripperSettings{}
+	newRT := func(cfg *tls.Config) (http.RoundTripper, error) {
+		return &http.Transport{TLSClientConfig: cfg}, nil
+	}
+	tlsRoundTripper, err := prometheusconfig.NewTLSRoundTripperWithContext(r.ctx, tlsConfig, settings, newRT)
+	if err != nil {
+		r.Log.Error(err, "failed to create tls round tripper")
+		return nil, err
+	}
+	roundTripper := prometheusconfig.NewAuthorizationCredentialsRoundTripper("Bearer", prometheusconfig.NewFileSecret(tokenPath), tlsRoundTripper)
+	// create prometheus client
+	prometheusClient, err := prometheusapi.NewClient(prometheusapi.Config{
+		Address:      prometheusURL,
+		RoundTripper: roundTripper,
+	})
+	if err != nil {
+		r.Log.Error(err, "failed to create prometheus client")
+		return nil, err
+	}
+	// scrape the metrics
+	scraper := prometheusv1.NewAPI(prometheusClient)
+	osdUsageQuery := "(ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))"
+	result, warnings, err := scraper.Query(context.Background(), osdUsageQuery, time.Now(), prometheusv1.WithTimeout(10*time.Second))
+	if err != nil {
+		r.Log.Error(err, "failed to query prometheus")
+		return nil, err
+	}
+
+	if len(warnings) > 0 {
+		r.Log.Info("prometheus query warnings", "warnings", warnings)
+	}
+
+	var vector model.Vector
+	if result.Type() == model.ValVector {
+		vector = result.(model.Vector)
+	}
+
+	return vector, nil
+}
+
+// update the sync map with the highest value of osd usage per device class
+func (r *StorageAutoscalerReconciler) updateSyncMap(metrics model.Vector) {
+	// create a map of device class to highest osd usage
+	deviceClassUsage := make(map[string]float64)
+	for _, osd := range metrics {
+		deviceClass := osd.Metric["device_class"]
+		if _, ok := deviceClassUsage[string(deviceClass)]; ok {
+			if float64(osd.Value) > deviceClassUsage[string(deviceClass)] {
+				deviceClassUsage[string(deviceClass)] = float64(osd.Value)
+			}
+		} else {
+			deviceClassUsage[string(deviceClass)] = float64(osd.Value)
+		}
+	}
+
+	// update the sync map
+	for deviceClass, usage := range deviceClassUsage {
+		r.SyncMap.Store(deviceClass, usage)
+	}
+}
+
+func (r *StorageAutoscalerReconciler) getStorageAutoScalerObjects() ([]struct {
```

Comment:
The function is called "get storage autoscaler objects", but the function is clearly doing a lot more than just that. Can this be broken down more?

Reply:
we have logs that clearly define whats its purpose 
`get resource objects from kube api that need scaling`

I don't think breaking it further will provide any more clarity, we can change the function name is there is any good suggestion